[AArch64] Dont inline streaming fn into non-streaming caller

sdesmalen-arm opened the pull request:
Without this change, the following test would fail to compile
with `-march=armv8-a+sme`:

```
  void func1(const svuint32_t *in, svuint32_t *out) {
    [&]() __arm_streaming { *out = *in; }();
  }
```

But in general, it's probably better never to inline
streaming functions into non-streaming functions, because
they will have been marked as 'streaming' for a reason
by the user.

sdesmalen-arm committed:
[AArch64] Dont inline streaming fn into non-streaming caller

Without this change, the following test would fail to compile
with `-march=armv8-a+sme`:

  void func1(const svuint32_t *in, svuint32_t *out) {
    [&]() __arm_streaming { *out = *in; }();
  }

But in general, it's probably better never to inline
streaming functions into non-streaming functions, because
they will have been marked as 'streaming' for a reason
by the user.

sdesmalen-arm committed:
Precommit test

gbossu commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -280,16 +280,22 @@ bool AArch64TTIImpl::areInlineCompatible(const Function *Caller,
+
+  // Cannot inline a streaming function into a non-streaming function,
+  // if the caller has no SVE.
+  if (CallAttrs.requiresSMChange() && !CallerBits.test(AArch64::FeatureSVE))
```
I'm wondering if there's other features we should test for? (Maybe not now, but I don't know what the future holds). Is there maybe a more generic way to check that the features of the callee are supported by the caller?

MacDue replied:
I guess in the general case this the features being compatible is checked by `(EffectiveCallerBits & EffectiveCalleeBits) == EffectiveCalleeBits`  (below this check). But streaming mode is not a target feature (just a function attribute), so is not included in this check.

MacDue reviewed (approved):
LGTM (with a minor comment, but I think this is okay as-is)

MacDue commented on llvm/test/Transforms/Inline/AArch64/sme-pstatesm-attrs.ll:
```diff
@@ -583,67 +583,67 @@ entry:
-; CHECK-LABEL: define void @streaming_caller_multiple_streaming_callees_dont_inline
-; CHECK-SAME: () #[[ATTR6]] {
-; CHECK-NEXT:    call void @streaming_caller_multiple_streaming_callees()
+define void @nonstreaming_caller_multiple_nonstreaming_callees_dont_inline() #0 "aarch64_pstate_sm_enabled" {
```
Is `aarch64_pstate_sm_enabled` intentional here? The function is `nonstreaming_caller`, so I'd guess not? Generally, I'd prefer that the precommit does not change any existing tests other than adding new cases (since it makes it harder to tell what's changed due to the LLVM change). If tests need tweaking, I'd be more obvious to see why as a commit after the LLVM change.

sdesmalen-arm replied:
> Generally, I'd prefer that the precommit does not change any existing tests other than adding new cases (since it makes it harder to tell what's changed)

I've rejigged the test changes a bit. Let me know if this makes it easier to review.

MacDue commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -270,6 +270,13 @@ bool AArch64TTIImpl::areInlineCompatible(const Function *Caller,
+  // into a non-streaming function. Assume it was marked as streaming
+  // for a reason.
+  if (CallAttrs.caller().hasNonStreamingInterfaceAndBody() &&
+      CallAttrs.callee().hasStreamingInterfaceOrBody())
```
This is specifically allowed (in a limited cases) in https://github.com/llvm/llvm-project/pull/68415 as an optimization (though come to think of it, there are cases where doing this could result in worse code, e.g., if the inlined call contains many calls that result in streaming mode changes).

sdesmalen-arm committed:
Fix tests

Because the behaviour now changed, I've had to modify some of the
other tests to test inlining with streaming mode changes the
other way around (streaming <- non-streaming instead of
non-streaming <- streaming)

MacDue reviewed (approved):
LGTM, thanks! Just a few nits on comments in the tests:

sdesmalen-arm committed:
Update comments in test

gbossu reviewed (approved):
LGTM
